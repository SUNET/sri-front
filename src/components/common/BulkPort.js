import React, { Component } from 'react';
// Common imports
import { withTranslation } from 'react-i18next';
import { Form } from 'react-bootstrap';
import { isBrowser } from 'react-device-detect';
import Dropdown from '../Dropdown';

import { CREATE } from '../../utils/constants';

import '../../style/bulkCreate.scss';

const titleByInputName = {
  type: 'general-forms/type',
  prefix: 'general-forms/bulk-port/prefix',
  offset: 'general-forms/bulk-port/offset',
};

const placeHolderByInputName = {
  type: 'general-forms/write-type',
  prefix: 'general-forms/bulk-port/prefix-placeholder',
  offset: 'general-forms/bulk-port/offset-placeholder',
  numberOfPorts: 'general-forms/bulk-port/Number-of-ports',
};

const errorByInputName = {
  type: 'general-forms/validations/required',
  prefix: 'general-forms/validations/required',
  offset: 'general-forms/validations/must-be-int',
  numberOfPorts: 'general-forms/validations/must-be-int',
};

function validRegularInput(value) {
  return !!value;
}

function validNumberInput(value) {
  const numberIntValue = Number.parseInt(value, 10);
  const isCorrectIntValue = Number.isInteger(numberIntValue);
  return isCorrectIntValue;
}

function validateField(inputName, value) {
  if (inputName === 'type' || inputName === 'prefix') {
    return validRegularInput(value);
  }
  return validNumberInput(value);
}

class BulkPort extends Component {
  constructor(props) {
    super(props);
    this.state = {
      type: { value: '', valid: true },
      prefix: { value: '', valid: true },
      offset: { value: 0, valid: true },
      numberOfPorts: { value: '', valid: true },
    };
  }

  createNewPortsList() {
    const { type, prefix, offset, numberOfPorts } = this.state;
    const portList = [];
    let index = parseInt(offset.value, 10);
    const maxNumberWithOffset = parseInt(offset.value, 10) + parseInt(numberOfPorts.value, 10);
    for (index; index < maxNumberWithOffset; index += 1) {
      portList.push({
        type: {
          name: type.value,
          value: type.value,
        },
        __typename: 'Port',
        name: `${prefix.value}${index}`,
        id: `port-to-create-${index}-${maxNumberWithOffset}`,
        status: CREATE,
      });
    }
    return portList;
  }

  handleCreatePorts() {
    const { handleBulkPortResponse } = this.props;

    const isValid = this.areValidFields();

    if (isValid) {
      handleBulkPortResponse(this.createNewPortsList());
    }
  }

  areValidFields() {
    const { type, prefix, offset, numberOfPorts } = this.state;
    const typeValid = validateField('type', type.value);
    const prefixValid = validateField('prefix', prefix.value);
    const offsetValid = validateField('offset', offset.value);
    const numberOfPortsValid = validateField('numberOfPorts', numberOfPorts.value);
    this.setState({
      type: { ...type, ...{ valid: typeValid } },
      prefix: { ...prefix, ...{ valid: prefixValid } },
      offset: { ...offset, ...{ valid: offsetValid } },
      numberOfPorts: { ...numberOfPorts, ...{ valid: numberOfPortsValid } },
    });
    const areValidAllFields = typeValid && prefixValid && offsetValid && numberOfPortsValid;
    return areValidAllFields;
  }

  renderInput(inputName, hasError) {
    const { t } = this.props;
    const stateValue = this.state[inputName];
    const { value } = stateValue;
    return (
      <Form.Group>
        <div className={`${hasError ? 'has-error' : ''}`}>
          <Form.Control
            type="text"
            value={value}
            placeholder={t(placeHolderByInputName[inputName])}
            onChange={(event) => {
              this.setState({
                [inputName]: {
                  ...stateValue,
                  ...{ value: event.target.value },
                },
              });
            }}
          />
          {hasError && <span>{t(errorByInputName[inputName])}</span>}
        </div>
      </Form.Group>
    );
  }

  renderRegularInput(inputName) {
    const { t } = this.props;
    const stateValue = this.state[inputName];
    const hasError = !stateValue.valid;

    return (
      <div className={`form-internal-block__section bulk-create__${inputName}-block`} key={`bulk-port-${inputName}`}>
        <div className="form-internal-block__section__title">{t(titleByInputName[inputName])}</div>
        <div className={'form-internal-block__section__content form-internal-block__section__content--edition-mode'}>
          {this.renderInput(inputName, hasError)}
        </div>
      </div>
    );
  }

  renderNumberPortsInput() {
    const { numberOfPorts } = this.state;
    const hasError = !numberOfPorts.valid;
    const fieldName = 'ports-number';
    return (
      <div className={`bulk-create__right-block__number-input bulk-create__${fieldName}-block`}>
        {this.renderInput('numberOfPorts', hasError)}
      </div>
    );
  }

  renderSelectTypeInput() {
    const { t } = this.props;
    const { type } = this.state;
    const hasError = !type.valid;
    const fieldName = 'type';
    return (
      <div className={`form-internal-block__section bulk-create__${fieldName}-block`} key={`bulk-port-${'type'}`}>
        <div className="form-internal-block__section__title">{t(titleByInputName['type'])}</div>
        <div className={'form-internal-block__section__content form-internal-block__section__content--edition-mode'}>
          <Form.Group>
            <div className={`select-input ${hasError ? 'has-error' : ''}`}>
              <Dropdown
                t={t}
                className={`${isBrowser ? 'auto' : 'xlg mw-100'}`}
                emptyLabel="Select type"
                type="port_types"
                name="bulk-port_type"
                onChange={(event) => {
                  this.setState({
                    type: {
                      ...type,
                      ...{ value: event.value, valid: validateField('type', event.value) },
                    },
                  });
                }}
              />
              {hasError && <span>{t(errorByInputName['type'])}</span>}
            </div>
          </Form.Group>
        </div>
      </div>
    );
  }

  renderCreatePortsButton() {
    const { t } = this.props;
    return (
      <div className="bulk-create__right-block__add-button">
        <button type="button" className="btn btn-add outline" onClick={(e) => this.handleCreatePorts(e)}>
          <span>{t('actions/add-new')}</span>
        </button>
      </div>
    );
  }

  render() {
    return (
      <div className="bulk-create">
        <div className="bulk-create__left-block form-internal-block">
          {this.renderSelectTypeInput()}
          {this.renderRegularInput('prefix')}
          {this.renderRegularInput('offset')}
        </div>
        <div className="d-flex bulk-create__right-block">
          {this.renderNumberPortsInput()}
          {this.renderCreatePortsButton()}
        </div>
      </div>
    );
  }
}

export default withTranslation()(BulkPort);
