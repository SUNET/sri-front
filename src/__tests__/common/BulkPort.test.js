import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import BulkPort from '../../components/common/BulkPort';

const mockMath = Object.create(global.Math);
mockMath.random = () => {
  return 1;
};
global.Math = mockMath;

let BulkPortWrapper;
let BulkPortComponent;
let BulkPortComponentInstance;

const defaultProps = {
  t: jest.fn((t) => t),
};

const CreateBulkPortWrapper = (props) => {
  const allProps = { ...defaultProps, ...props };
  return (
    <div>
      <BulkPort {...allProps} />
    </div>
  );
};

const mountComponentWrapper = (props) => mount(<CreateBulkPortWrapper {...props} />);

beforeEach(() => {
  jest.clearAllMocks();
  BulkPortWrapper = mountComponentWrapper();
  BulkPortComponent = BulkPortWrapper.find('BulkPort');
  BulkPortComponentInstance = BulkPortComponent.instance();
});
afterEach(() => {
  BulkPortWrapper.unmount();
});

describe('BulkPort Component', () => {
  it('Renders correctly', () => {
    expect(toJson(BulkPortWrapper)).toMatchSnapshot();
  });
});

describe('Add Events', () => {
  it('Validations Errors appears when click in ADD NEW with inputs empty', () => {
    const addNewButton = BulkPortComponent.find('.bulk-create__right-block__add-button button');
    addNewButton.simulate('click');
    expect(BulkPortComponent.update().find('.bulk-create__type-block .select-input.has-error')).toHaveLength(1);
    expect(BulkPortComponent.update().find('.bulk-create__prefix-block .has-error')).toHaveLength(1);
    expect(BulkPortComponent.update().find('.bulk-create__ports-number-block .has-error')).toHaveLength(1);
  });
  // it('When the fields are correct the list of ports is sent ', () => {
  //   BulkPortComponent.find('.bulk-create__type-block .select-input select').simulate('change', {target: { value: 'CEE' }});
  //   BulkPortComponent.find('.bulk-create__prefix-block input').simulate('change', { target: { value: 'prefix-test-' } });
  //   BulkPortComponent.find('.bulk-create__ports-number-block input').simulate('change', { target: { value: 5 } });
  //   BulkPortComponent.find('.bulk-create__right-block__add-button button').simulate('click');
  //   expect(BulkPortComponent.find('.bulk-create__type-block .select-input select')).toHaveLength(1);
  // });
  it('Filled from and to inputs', () => {
  });

  it('Click in reset dates ', () => {
  });
});
