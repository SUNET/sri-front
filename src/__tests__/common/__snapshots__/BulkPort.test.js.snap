// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkPort Component Renders correctly 1`] = `
<CreateBulkPortWrapper>
  <div>
    <withI18nextTranslation(BulkPort)
      t={[MockFunction]}
    >
      <BulkPort
        i18n={Object {}}
        t={[Function]}
        tReady={false}
      >
        <div
          className="bulk-create"
        >
          <div
            className="bulk-create__left-block form-internal-block"
          >
            <div
              className="form-internal-block__section bulk-create__type-block"
              key="bulk-port-type"
            >
              <div
                className="form-internal-block__section__title"
              >
                general-forms/type
              </div>
              <div
                className="form-internal-block__section__content form-internal-block__section__content--edition-mode"
              >
                <FormGroup>
                  <div
                    className="form-group"
                  >
                    <div
                      className="select-input "
                    >
                      <Dropdown
                        className="auto"
                        emptyLabel="Select type"
                        name="bulk-port_type"
                        onChange={[Function]}
                        type="port_types"
                      >
                        <FormGroup
                          className="d-inline"
                        >
                          <div
                            className="d-inline form-group"
                          >
                            <ReactRelayQueryRenderer
                              environment={"RelayModernEnvironment()"}
                              query={
                                Object {
                                  "fragment": Object {
                                    "argumentDefinitions": Array [
                                      Object {
                                        "defaultValue": null,
                                        "kind": "LocalArgument",
                                        "name": "name",
                                        "type": "String!",
                                      },
                                    ],
                                    "kind": "Fragment",
                                    "metadata": null,
                                    "name": "DropdownQuery",
                                    "selections": Array [
                                      Object {
                                        "alias": null,
                                        "args": Array [
                                          Object {
                                            "kind": "Variable",
                                            "name": "name",
                                            "variableName": "name",
                                          },
                                        ],
                                        "concreteType": "Choice",
                                        "kind": "LinkedField",
                                        "name": "getChoicesForDropdown",
                                        "plural": true,
                                        "selections": Array [
                                          Object {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "name",
                                            "storageKey": null,
                                          },
                                          Object {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "value",
                                            "storageKey": null,
                                          },
                                        ],
                                        "storageKey": null,
                                      },
                                    ],
                                    "type": "Query",
                                  },
                                  "hash": "a084763b46e47fd40d438714d89c94e1",
                                  "kind": "Request",
                                  "operation": Object {
                                    "argumentDefinitions": Array [
                                      Object {
                                        "defaultValue": null,
                                        "kind": "LocalArgument",
                                        "name": "name",
                                        "type": "String!",
                                      },
                                    ],
                                    "kind": "Operation",
                                    "name": "DropdownQuery",
                                    "selections": Array [
                                      Object {
                                        "alias": null,
                                        "args": Array [
                                          Object {
                                            "kind": "Variable",
                                            "name": "name",
                                            "variableName": "name",
                                          },
                                        ],
                                        "concreteType": "Choice",
                                        "kind": "LinkedField",
                                        "name": "getChoicesForDropdown",
                                        "plural": true,
                                        "selections": Array [
                                          Object {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "name",
                                            "storageKey": null,
                                          },
                                          Object {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "value",
                                            "storageKey": null,
                                          },
                                          Object {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "id",
                                            "storageKey": null,
                                          },
                                        ],
                                        "storageKey": null,
                                      },
                                    ],
                                  },
                                  "params": Object {
                                    "id": null,
                                    "metadata": Object {},
                                    "name": "DropdownQuery",
                                    "operationKind": "query",
                                    "text": "query DropdownQuery(
  $name: String!
) {
  getChoicesForDropdown(name: $name) {
    name
    value
    id
  }
}
",
                                  },
                                }
                              }
                              render={[Function]}
                              variables={
                                Object {
                                  "name": "port_types",
                                }
                              }
                            >
                              <select
                                className="auto"
                              >
                                <option
                                  value=""
                                >
                                  Select type
                                </option>
                              </select>
                            </ReactRelayQueryRenderer>
                          </div>
                        </FormGroup>
                      </Dropdown>
                    </div>
                  </div>
                </FormGroup>
              </div>
            </div>
            <div
              className="form-internal-block__section bulk-create__prefix-block"
              key="bulk-port-prefix"
            >
              <div
                className="form-internal-block__section__title"
              >
                general-forms/bulk-port/prefix
              </div>
              <div
                className="form-internal-block__section__content form-internal-block__section__content--edition-mode"
              >
                <FormGroup>
                  <div
                    className="form-group"
                  >
                    <div
                      className=""
                    >
                      <FormControl
                        onChange={[Function]}
                        placeholder="general-forms/bulk-port/prefix-placeholder"
                        type="text"
                        value=""
                      >
                        <input
                          className="form-control"
                          onChange={[Function]}
                          placeholder="general-forms/bulk-port/prefix-placeholder"
                          type="text"
                          value=""
                        />
                      </FormControl>
                    </div>
                  </div>
                </FormGroup>
              </div>
            </div>
            <div
              className="form-internal-block__section bulk-create__offset-block"
              key="bulk-port-offset"
            >
              <div
                className="form-internal-block__section__title"
              >
                general-forms/bulk-port/offset
              </div>
              <div
                className="form-internal-block__section__content form-internal-block__section__content--edition-mode"
              >
                <FormGroup>
                  <div
                    className="form-group"
                  >
                    <div
                      className=""
                    >
                      <FormControl
                        onChange={[Function]}
                        placeholder="general-forms/bulk-port/offset-placeholder"
                        type="text"
                        value={0}
                      >
                        <input
                          className="form-control"
                          onChange={[Function]}
                          placeholder="general-forms/bulk-port/offset-placeholder"
                          type="text"
                          value={0}
                        />
                      </FormControl>
                    </div>
                  </div>
                </FormGroup>
              </div>
            </div>
          </div>
          <div
            className="d-flex bulk-create__right-block"
          >
            <div
              className="bulk-create__right-block__number-input bulk-create__ports-number-block"
            >
              <FormGroup>
                <div
                  className="form-group"
                >
                  <div
                    className=""
                  >
                    <FormControl
                      onChange={[Function]}
                      placeholder="general-forms/bulk-port/Number-of-ports"
                      type="text"
                      value=""
                    >
                      <input
                        className="form-control"
                        onChange={[Function]}
                        placeholder="general-forms/bulk-port/Number-of-ports"
                        type="text"
                        value=""
                      />
                    </FormControl>
                  </div>
                </div>
              </FormGroup>
            </div>
            <div
              className="bulk-create__right-block__add-button"
            >
              <button
                className="btn btn-add outline"
                onClick={[Function]}
                type="button"
              >
                <span>
                  actions/add-new
                </span>
              </button>
            </div>
          </div>
        </div>
      </BulkPort>
    </withI18nextTranslation(BulkPort)>
  </div>
</CreateBulkPortWrapper>
`;
